File: 2018/Day_01/Python/solution.py
"""Advent of Code 2017 Day 01 - Chronal Calibration."""


def input_per_line(file: str):
    """This is for when each line is an input to the puzzle. The newline character is stripped."""
    with open(file, 'r') as input_file:
        return [eval(line.rstrip()) for line in input_file.readlines()]


def find_repeating_frequency(numbers: list[int]) -> int:
    """Go through the list over and over, summing one by one and looking for the first repeated sum.
    Then return that number.
    """
    seen_frequencies = set()
    current_frequency = 0
    searching_for_frequencies_repeated = True
    while searching_for_frequencies_repeated:
        for number in numbers:
            current_frequency += number
            if current_frequency in seen_frequencies:
                return current_frequency
            seen_frequencies.add(current_frequency)


if __name__ == "__main__":
    frequencies = input_per_line("../input.txt")
    part_one_sum = sum(frequencies)
    print(f"The resulting frequency is {part_one_sum}")
    part_two = find_repeating_frequency(frequencies)
    print(f"The first frequency reached twice is {part_two}")

File: 2018/Day_01/Go/solution.go
// Advent of Code 2018 Day 01 - Chronal Calibration
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

// MultipleLinesNumbers returns the input file if it contains multiple lines of text which are numbers
func MultipleLinesNumbers(fileName string) ([]int, error) {
	inputSlice := make([]int, 0)
	file, err := os.Open(fileName)
	if err != nil {
		return inputSlice, err
	}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		thisNumber, _ := strconv.Atoi(scanner.Text())
		inputSlice = append(inputSlice, thisNumber)
	}
	err = file.Close()
	if err != nil {
		return inputSlice, err
	}
	return inputSlice, nil
}

// totalFrequencies sums up the frequencies in the slice and returns the total
func totalFrequencies(frequencySlice []int) int {
	total := 0
	for _, number := range frequencySlice {
		total += number
	}
	return total
}

// findRepeatedFrequency works like totalFrequencies except it stops when it gets to one it's seen before
func findRepeatedFrequency(frequencySlice []int) int {
	fakeSet := map[int]bool{} // this will function as a set since Go doesn't have them built in
	total := 0
	for {
		for _, number := range frequencySlice {
			total += number
			if fakeSet[total] {
				return total
			} else {
				fakeSet[total] = true
			}
		}
	}
	return 0
}

func main() {
	frequencyChanges, err := MultipleLinesNumbers("../input.txt")
	if err != nil {
		print(err)
	}
	partOne := totalFrequencies(frequencyChanges)
	fmt.Printf("The sum of the frequencies is %v\n", partOne)
	partTwo := findRepeatedFrequency(frequencyChanges)
	fmt.Printf("The first repeated frequency is %v\n", partTwo)
}
